gboard_autopsy.py

```diff
@@ -0,0 +1,123 @@
+        return "Extracts Android GBoard app data."
+
+    def getModuleVersionNumber(self):
+        return "1.0"
+
+    def isDataSourceIngestModuleFactory(self):
+        return True
+
+    def createDataSourceIngestModule(self, ingestOptions):
+        return GboardDataSourceIngestModule()
+
+class GboardDataSourceIngestModule(DataSourceIngestModule):
+
+    _logger = Logger.getLogger(GboardDataSourceIngestModuleFactory.moduleName)
+
+    GBOARD_CLIPBOARD_ARTIFACT = 'GBOARD_CLIPBOARD_OBJECT'
+    GBOARD_PACKAGE_NAME = 'com.google.android.inputmethod.latin'
+
+    def log(self, level, msg):
+        self._logger.logp(level, self.__class__.__name__, inspect.stack()[1][3], msg)
+
+    def __init__(self):
+        self.context = None
+
+    def startUp(self, context):
+        """Where module setup and configuration is done
+
+        Args:
+        context (IngestJobContext): Context of the ingest module
+        """
+
+        self.log(Level.INFO, "GBoard Data Source Ingest Module is starting up...")
+
+        # setup the current context
+        self.context = context
+
+        # check if executable exists
+        module_path = os.path.dirname(os.path.abspath(__file__))
+        exe_ext = '.exe' if os.name == 'nt' else ''
+        path_to_exe = os.path.join(module_path, 'gboard-forensics' + exe_ext)
+
+        if not os.path.exists(path_to_exe):
+            raise IngestModuleException('Executable file not found!')
+
+        try:
+            current_case = Case.getCurrentCaseThrows()
+
+            # create artifacts
+            self.clipboard_type = self.createCustomArtifactType(current_case, self.GBOARD_CLIPBOARD_ARTIFACT, 'Gboard Clipboard')
+        except NoCurrentCaseException as ex:
+            self.log(Level.WARNING, "No case currently open. " + ex)
+
+    def process(self, data_source, progress_bar):
+        """Where the analysis is done
+
+        Args:
+        dataSource (Content): autopsy case data source
+        progressBar (DataSourceIngestModuleProgress): progress bar related to this analysis
+        """
+
+        progress_bar.switchToIndeterminate()
+        self.log(Level.INFO, "GBoard Data Source Ingest Module is analysing...")
+
+        try:
+            current_case = Case.getCurrentCaseThrows()
+            services = current_case.getServices()
+
+            blackboard = services.getBlackboard()
+            file_manager = services.getFileManager()
+
+            files = file_manager.findFiles('%', self.GBOARD_PACKAGE_NAME + '/files/clipboard_image')
+            for file in files:
+                self.log(Level.INFO, file.toString())
+                artifact = file.newArtifact(self.clipboard_type.getTypeID())
+
+                try:
+                    # index the artifact for keyword search
+                    blackboard.indexArtifact(artifact)
+                except Blackboard.BlackboardException as ex:
+                    self.log(Level.SEVERE, "Error indexing artifact " + artifact.getDisplayName())
```

Comment:
This shouldn't be `SEVERE`. `SEVERE` is the highest value and is a level that indicates the program won't be able to run normally. And because Autopsy doesn't have a level that higher, `SEVERE` should be used as a `FATAL/CRITICAL` level. From the official oracle's wiki `... which will prevent normal program execution`. Autopsy uses the same meanings as Java's logging util.

https://docs.oracle.com/javase/6/docs/api/java/util/logging/Level.html#SEVERE

Unless you want to terminate the ongoing action, I suggest switching to `WARNING` because you're technically handling the error yourself.

Reply:
I would say that should be a Level.ERROR or something, but I think this is correct. Also, the autopsy example uses the same Level here: https://github.com/sleuthkit/autopsy/blob/develop/pythonExamples/dataSourceIngestModule.py#L149 . If you feel more confortable I can change it, but I don't know the drawbacks it can come up with. Maybe autopsy uses this levels to show some severe errors on their logging system, idk.

Reply:
Well, Autopsy should have more logging levels. I don't really like the levels in Java but at least they have more, and if Autopsy is using those same level meanings for their own, then they could add more. I don't see why they chose `SEVERE` here as they are continuing with the program execution. I'm guessing this would only affect the final visualization product going, once again, towards a `WARNING` logging level. I think `ERROR` would be fine if they had it. The same inconsistency is done below in **line 120**. They use `INFO` for that error, but in other parts of the example, they use `WARNING`. I won't force the use of other levels, however, I would like some consistency on the usage, which I believe isn't happening in the example that you followed.